File: ios/AlphaCityStampTour/AlphaCityStampTour/ViewModel/StoreRegisterViewModel.swift
//
//  StoreRegisterViewModel.swift
//  AlphaCityStampTour
//

import Foundation

@MainActor
final class StoreRegisterViewModel: ObservableObject {
    @Published var isSubmitting = false
    @Published var submitResult: SubmitResult? = nil
    @Published var isUploadingImage = false
    @Published var uploadedImageUrl: String? = nil

    private let repository = StoreRepository.shared
    private let client = APIClient.shared

    enum SubmitResult {
        case success
        case error(String)
    }

    // MARK: - 대표 이미지 업로드

    func uploadImage(_ data: Data) {
        guard !isUploadingImage else { return }
        isUploadingImage = true

        Task {
            do {
                uploadedImageUrl = try await client.uploadImage(data, filename: "store_\(Int(Date().timeIntervalSince1970)).jpg")
            } catch {
                print("[StoreRegisterVM] 이미지 업로드 실패: \(error)")
                uploadedImageUrl = nil
            }
            isUploadingImage = false
        }
    }

    func registerStore(
        name: String,
        category: String,
        ownerName: String,
        phone: String,
        address: String,
        addressDetail: String,
        description: String,
        storeCode: String,
        operatingDays: String,
        openTime: String,
        closeTime: String
    ) {
        guard !isSubmitting else { return }
        isSubmitting = true

        Task {
            let request = StoreRegisterRequest(
                name: name,
                category: category,
                ownerName: ownerName,
                phone: phone,
                address: address.isEmpty ? nil : address,
                addressDetail: addressDetail.isEmpty ? nil : addressDetail,
                description: description.isEmpty ? nil : description,
                imageUrl: uploadedImageUrl,
                storeCode: storeCode.isEmpty ? nil : storeCode,
                operatingDays: operatingDays.isEmpty ? nil : operatingDays,
                openTime: openTime.isEmpty ? nil : openTime,
                closeTime: closeTime.isEmpty ? nil : closeTime
            )

            do {
                _ = try await repository.registerStore(request)
                submitResult = .success
            } catch {
                print("[StoreRegisterVM] 등록 실패: \(error)")
                submitResult = .error(error.localizedDescription)
            }

            isSubmitting = false
        }
    }

    func clearResult() {
        submitResult = nil
    }
}

File: android/app/src/main/java/com/alphacity/stamptour/viewmodel/StoreRegisterViewModel.kt
package com.alphacity.stamptour.viewmodel

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.alphacity.stamptour.network.dto.StoreRegisterRequest
import com.alphacity.stamptour.repository.StoreRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class StoreRegisterViewModel @Inject constructor(
    private val storeRepository: StoreRepository,
) : ViewModel() {

    private val _isSubmitting = MutableStateFlow(false)
    val isSubmitting: StateFlow<Boolean> = _isSubmitting

    private val _submitResult = MutableStateFlow<SubmitResult?>(null)
    val submitResult: StateFlow<SubmitResult?> = _submitResult

    fun registerStore(
        name: String,
        category: String,
        ownerName: String,
        phone: String,
        address: String,
        addressDetail: String,
        description: String,
        storeCode: String,
        operatingDays: String,
        openTime: String,
        closeTime: String,
    ) {
        if (_isSubmitting.value) return
        _isSubmitting.value = true

        viewModelScope.launch {
            val request = StoreRegisterRequest(
                name = name,
                category = category,
                ownerName = ownerName,
                phone = phone,
                address = address.ifBlank { null },
                addressDetail = addressDetail.ifBlank { null },
                description = description.ifBlank { null },
                storeCode = storeCode.ifBlank { null },
                operatingDays = operatingDays.ifBlank { null },
                openTime = openTime.ifBlank { null },
                closeTime = closeTime.ifBlank { null },
            )

            storeRepository.registerStore(request)
                .onSuccess {
                    _submitResult.value = SubmitResult.Success
                }
                .onFailure { e ->
                    Log.e("StoreRegisterVM", "등록 실패", e)
                    _submitResult.value = SubmitResult.Error(e.message ?: "등록에 실패했습니다")
                }

            _isSubmitting.value = false
        }
    }

    fun clearResult() {
        _submitResult.value = null
    }

    sealed class SubmitResult {
        data object Success : SubmitResult()
        data class Error(val message: String) : SubmitResult()
    }
}
